cpp/src/arrow/compute/key_map_internal.h

```diff
@@ -81,31 +81,70 @@ class ARROW_EXPORT SwissTable {
 
   void num_inserted(uint32_t i) { num_inserted_ = i; }
 
-  uint8_t* blocks() const { return blocks_->mutable_data(); }
-
   uint32_t* hashes() const {
     return reinterpret_cast<uint32_t*>(hashes_->mutable_data());
   }
 
   /// \brief Extract group id for a given slot in a given block.
   ///
-  inline uint64_t extract_group_id(const uint8_t* block_ptr, int slot,
-                                   uint64_t group_id_mask) const;
+  static uint32_t extract_group_id(const uint8_t* block_ptr, int local_slot,
+                                   int num_group_id_bits) {
+    // Extract group id using aligned 32-bit read.
```

Comment:
Why not use `SafeLoad` as in other places already?
(also, since this is non-trivial, factoring out the loading of a group id could go into a dedicated inline function)

Reply:
The original code uses aligned read + masking so I'm following it, possibly for performance sake I guess?

If `SafeLoad` is preferred (i.e., it doesn't hurt performance), then yes it is possible to factor this piece of code out.

Reply:
> I think we should at least keep 2) as is because it makes perfect sense. 1) and 3) can be unified, either aligned or unaligned.

Agreed. Feel free to choose whatever approach you prefer!

Reply:
Addressed. Thank you for the suggestion.

This PR is good for review again.